ledger/state.go

```diff
@@ -979,19 +987,202 @@ func (ls *LedgerState) forgeBlock() {
+					for _, mempoolTx := range mempoolTxs {
+						// Marshal each transaction to get its size
+						txCbor, err := cbor.Encode(mempoolTx)
+						if err != nil {
+							ls.config.Logger.Debug(
+								"failed to marshal transaction, skipping",
+								"component", "ledger",
+								"error", err,
+							)
+							continue
+						}
+						txSize := len(txCbor)
+						// Check MaxTxSize limit with safe conversion
+						var txSizeUint uint
+						if txSize >= 0 {
+							txSizeUint = uint(txSize)
+						} else {
+							// If txSize is negative, use max value
+							txSizeUint = ^uint(0)
+						}
+						if txSizeUint > maxTxSize {
+							ls.config.Logger.Debug(
+								"skipping transaction - exceeds MaxTxSize",
+								"component", "ledger",
+								"tx_size", txSize,
+								"max_tx_size", maxTxSize,
+							)
+							continue
+						}
+
+						// Check MaxBlockSize limit and added afe conversion to prevent integer overflow
+						var totalBlockSize uint
+						if blockSize >= 0 && txSize >= 0 {
+							// Check for overflow by comparing with max int
+							maxInt := 1<<31 - 1
+							if blockSize <= maxInt && txSize <= maxInt && blockSize+txSize <= maxInt {
+								// Safe conversion
+								totalBlockSize = uint(uint64(blockSize) + uint64(txSize))
+							} else {
+								// use max value in case of overflow
+								totalBlockSize = ^uint(0)
+							}
+						} else {
+							// use max value in case of negative
+							totalBlockSize = ^uint(0)
+						}
+						if totalBlockSize > maxBlockSize {
+							ls.config.Logger.Debug(
+								"block size limit reached",
+								"component", "ledger",
+								"current_size", blockSize,
+								"tx_size", txSize,
+								"max_block_size", maxBlockSize,
+							)
+							break
+						}
+						// Safe conversion to prevent integer overflow
+						var estimatedMemory, estimatedSteps uint64
+						if txSize >= 0 && txSize <= int(^uint(0)>>1) {
+							// Safe conversion
+							txSizeUint64 := uint64(txSize)
+							estimatedMemory = txSizeUint64 * 2
+							estimatedSteps = txSizeUint64 * 10
+						} else {
+							estimatedMemory = 0
+							estimatedSteps = 0
+						}
+						estimatedTxExUnits := lcommon.ExUnits{
+							Memory: estimatedMemory,
+							Steps:  estimatedSteps,
+						}
+
+						// Check MaxExUnits limit
+						if totalExUnits.Memory+estimatedTxExUnits.Memory > maxExUnits.Memory ||
+							totalExUnits.Steps+estimatedTxExUnits.Steps > maxExUnits.Steps {
+							ls.config.Logger.Debug(
+								"ex units limit reached",
+								"component", "ledger",
+								"current_memory", totalExUnits.Memory,
+								"current_steps", totalExUnits.Steps,
+								"tx_memory", estimatedTxExUnits.Memory,
+								"tx_steps", estimatedTxExUnits.Steps,
+								"max_memory", maxExUnits.Memory,
+								"max_steps", maxExUnits.Steps,
+							)
+							break
+						}
+
+						// Create Conway transaction structures from mempool transaction
+						var fullTx conway.ConwayTransaction
```

Comment:
Move the decode above the ExUnits calculation above so you can use the values straight from the transaction.